src/qibolab/transpilers/placer.py

```diff
@@ -197,33 +202,44 @@ def cost(graph, circuit_repr):
         return total_len - allowed
 
 
-# TODO: requires block decomposition
 class Backpropagation(Placer):
     """
-    Place qubits based on the algorithm proposed in
-    https://doi.org/10.48550/arXiv.1809.02573
+    Place qubits based on the algorithm proposed in https://doi.org/10.48550/arXiv.1809.02573.
+    Works with ShortestPaths routing.
+
+    Attributes:
+        connectivity (networkx.Graph): chip connectivity.
+        routing_algorithm (qibolab.transpilers.routing.Transpiler): routing algorithm.
+        iterations (int): number of executions of the forward and backward routing steps.
     """
 
-    def __init__(self, connectivity, routing_algorithm, iterations=1, max_lookahead_gates=None):
+    def __init__(self, connectivity, routing_algorithm, iterations=1):
         self.connectivity = connectivity
-        self._routing = routing_algorithm
-        self._iterations = iterations
-        self._max_gates = max_lookahead_gates
+        self.routing_algorithm = routing_algorithm
+        self.iterations = iterations
 
-    def __call__(self, circuit):
-        # Start with trivial placement
-        self._circuit_repr = create_circuit_repr(circuit)
-        initial_placement = dict(zip(list("q" + str(i) for i in range(circuit.nqubits)), range(circuit.nqubits)))
-        for _ in range(self._iterations):
-            final_placement = self.forward_step(initial_placement)
-            initial_placement = self.backward_step(final_placement)
-        return initial_placement
+    def __call__(self, circuit: Circuit):
+        """Find the initial layout of the given circuit using subgraph isomorphism.
 
-    # TODO: requires block circuit
-    def forward_step(self, initial_placement):
+        Args:
+            circuit (qibo.models.Circuit): circuit to be transpiled.
+        """
+
+        initial_placer = Trivial(self.connectivity)
+        initial_placement = initial_placer(circuit=circuit)
+        reversed_circuit = circuit.invert()
+        for _ in range(self.iterations):
+            final_placement = self.routing_step(initial_placement, circuit)
+            initial_placement = self.routing_step(final_placement, reversed_circuit)
```

Comment:
All of this has a bunch of assumptions hard-coded:
- you start from the beginning
- every time that you go forward, you do fully, and then
- you also go backward right after

I would just play more at circuit level: just make a single `routing_step` on a single `Circuit`, and then provide some other functions to manipulate the circuit you pass:
- you can `.invert()` and concatenate (as done here)
- start from a different point
- take only gates that are close to the beginning
- and so on...

Essentially, you could just accept a function that takes your original circuit, and construct a derived circuit out of it, and then using the output circuit for the `routing_step`.
Of course, this output circuit should respect some conditions (like ending at the starting point), but you can describe the meaning and leave this up to the user.
You could provide the current function (back-and-forth-n-times) as a default value for the user manipulation function :)

Reply:
This is just the sketch of the code. i want to concentrate on this function in the next pr

Reply:
Fine, I'm converting this to an issue then :)